File: 451. Sort Characters By Frequency/frequencySort.js
/**
 * @param {string} s
 * @return {string}
 */
var frequencySort = function (s) {
    let freq = {}
    let ans = ''

    for (let chr of s) {                                          //O(n)
        freq[chr] = (freq[chr] || 0) + 1
    }

    let sort = Object.entries(freq).sort((a, b) => b[1] - a[1])      //O(n log n)

    for (let [str, cnt] of sort) {                                //O(n)
        ans += str.repeat(cnt)
    }

    return ans;
};

File: 451. Sort Characters By Frequency/frequencySort.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
#include <queue>

using namespace std;

class Solution {
    public:
        string frequencySort(string s) {
            unordered_map<char, int> freq;
            string ans = "";
            for (char ch : s) {
                freq[ch]++;
            }
    
            //Use a max-heap priority_queue to sort by frequency
            priority_queue<pair<int, char>> pq;
            for (const auto& [ch, count] : freq) {
                pq.push({count, ch});
            }
    
            while(!pq.empty()){
                auto [cnt, value] = pq.top();
                pq.pop();
                for(int i=0; i<cnt; i++){
                    ans+=value;
                }
            }
    
            return ans;
        }
    };